staging/src/k8s.io/client-go/tools/clientcmd/loader.go

```diff
@@ -245,25 +244,72 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 	}
 
 	// first merge all of our maps
-	mapConfig := clientcmdapi.NewConfig()
+	var config *clientcmdapi.Config
 
 	for _, kubeconfig := range kubeconfigs {
-		mergo.Merge(mapConfig, kubeconfig, mergo.WithOverride)
-	}
+		if config == nil {
+			config = kubeconfig.DeepCopy()
+			continue
+		}
+
+		// Merge struct fields. Last non-empty wins.
+		if len(kubeconfig.APIVersion) > 0 || len(kubeconfig.Kind) > 0 {
+			config.APIVersion = kubeconfig.APIVersion
+			config.Kind = kubeconfig.Kind
+		}
+		if len(kubeconfig.CurrentContext) > 0 {
+			config.CurrentContext = kubeconfig.CurrentContext
+		}
+		if kubeconfig.Preferences.Colors {
+			config.Preferences.Colors = kubeconfig.Preferences.Colors
+		}
 
-	// merge all of the struct values in the reverse order so that priority is given correctly
-	// errors are not added to the list the second time
-	nonMapConfig := clientcmdapi.NewConfig()
-	for i := len(kubeconfigs) - 1; i >= 0; i-- {
-		kubeconfig := kubeconfigs[i]
-		mergo.Merge(nonMapConfig, kubeconfig, mergo.WithOverride)
+		// Merge map fields. First wins.
+		for k, v := range kubeconfig.Clusters {
+			if _, exists := config.Clusters[k]; !exists {
+				if config.Clusters == nil {
+					config.Clusters = map[string]*clientcmdapi.Cluster{}
+				}
+				config.Clusters[k] = v
+			}
+		}
+		for k, v := range kubeconfig.Contexts {
+			if _, exists := config.Contexts[k]; !exists {
+				if config.Contexts == nil {
+					config.Contexts = map[string]*clientcmdapi.Context{}
+				}
+				config.Contexts[k] = v
+			}
+		}
+		for k, v := range kubeconfig.AuthInfos {
+			if _, exists := config.AuthInfos[k]; !exists {
+				if config.AuthInfos == nil {
+					config.AuthInfos = map[string]*clientcmdapi.AuthInfo{}
+				}
+				config.AuthInfos[k] = v
+			}
+		}
+		for k, v := range kubeconfig.Extensions {
+			if _, exists := config.Extensions[k]; !exists {
+				if config.Extensions == nil {
+					config.Extensions = map[string]runtime.Object{}
+				}
+				config.Extensions[k] = v
+			}
+		}
+		for k, v := range kubeconfig.Preferences.Extensions {
+			if _, exists := config.Preferences.Extensions[k]; !exists {
+				if config.Preferences.Extensions == nil {
+					config.Preferences.Extensions = map[string]runtime.Object{}
+				}
+				config.Preferences.Extensions[k] = v
+			}
+		}
```

Comment:
do the existing unit tests exercise merge precedence for duplicate keys in extensions and preferences.extensions? I want to make sure we're not changing this behavior

Reply:
Yes @liggitt that's how we found the problem. https://github.com/kubernetes/kubernetes/blob/c6b5191c37f939d2d61e76de222a96ae5f5d9558/staging/src/k8s.io/client-go/tools/clientcmd/client_config_test.go#L34

also see https://github.com/kubernetes/kubernetes/issues/107499#issue-1100542382